ChopChop/Storage/AppDatabase.swift

```diff
@@ -329,45 +355,89 @@ extension AppDatabase {
 extension AppDatabase {
     func saveRecipe(_ recipe: inout RecipeRecord) throws {
         var ingredients: [RecipeIngredientRecord] = []
+        var graph = RecipeStepGraphRecord()
         var steps: [RecipeStepRecord] = []
-
-        try saveRecipe(&recipe, ingredients: &ingredients, steps: &steps)
+        var edges: [RecipeStepEdgeRecord] = []
+        var endPoints: [(source: RecipeStepRecord, destination: RecipeStepRecord)] = []
+
+        try saveRecipe(
+            &recipe,
+            ingredients: &ingredients,
+            graph: &graph,
+            steps: &steps,
+            edges: &edges,
+            endPoints: endPoints)
     }
 
     func saveRecipe(_ recipe: inout RecipeRecord,
                     ingredients: inout [RecipeIngredientRecord],
-                    steps: inout [RecipeStepRecord]) throws {
+                    graph: inout RecipeStepGraphRecord,
+                    steps: inout [RecipeStepRecord],
+                    edges: inout [RecipeStepEdgeRecord],
+                    endPoints: [(source: RecipeStepRecord, destination: RecipeStepRecord)]) throws {
         try dbWriter.write { db in
             try recipe.save(db)
 
-            let recipeIds = ingredients.compactMap { $0.recipeId } + steps.compactMap { $0.recipeId }
+            let recipeIds = ingredients.compactMap { $0.recipeId } + ((graph.recipeId == nil) ? [] : [graph.recipeId])
```

Comment:
Nit but you can probably do `let recipeIds = ingredients.compactMap { $0.recipeId } + [graph.recipeId].compactMap { $0 }`.